aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,764 @@
+
+    use crate::{
+        aggregator::queue::{queued_lifo, LIFORequestQueue, Metrics},
+        metrics::test_util::InMemoryMetricsInfrastructure,
+    };
+
+    /// Some tests busy loop waiting for a condition to become true. Avoid hanging broken tests
+    /// indefinitely by wrapping those tests in a timeout.
+    const TEST_TIMEOUT: Duration = Duration::from_secs(15);
+
+    async fn get_outstanding_requests_gauge(
+        metrics: &InMemoryMetricsInfrastructure,
+        meter_prefix: &str,
+    ) -> Option<usize> {
+        Some(
+            metrics
+                .collect()
+                .await
+                // The metric may not be immediately available when we need it, so return an Option
+                // instead of unwrapping.
+                .get(&Metrics::get_outstanding_requests_name(meter_prefix))?
+                .data
+                .as_any()
+                .downcast_ref::<Gauge<u64>>()
+                .unwrap()
+                .data_points[0]
+                .value
+                .try_into()
+                .unwrap(),
+        )
+    }
+
+    struct HangingHandler {
+        unhang: Arc<Notify>,
+    }
+
+    #[async_trait]
+    impl Handler for HangingHandler {
+        async fn run(&self, conn: trillium::Conn) -> trillium::Conn {
+            let _ = self.unhang.notified().await;
+            conn.ok("hello")
+        }
+    }
+
+    async fn fill_queue(
+        handler: Arc<impl Handler>,
+        concurrency: u32,
+        depth: usize,
+        metrics: &InMemoryMetricsInfrastructure,
+        meter_prefix: &str,
+    ) -> Vec<JoinHandle<()>> {
+        debug!("filling queue");
+
+        let mut requests = Vec::new();
+
+        let backoff = ExponentialBackoff {
+            initial_interval: Duration::from_nanos(1),
+            max_interval: Duration::from_nanos(30),
+            multiplier: 2.0,
+            ..Default::default()
+        };
+
+        debug!("spawning requests");
+        for _ in 0..(concurrency as usize + depth) {
+            let handler = Arc::clone(&handler);
+            let backoff = backoff.clone();
+            requests.push(tokio::spawn({
+                async move {
+                    retry(backoff, || {
+                        let handler = Arc::clone(&handler);
+                        async move {
+                            let request = get("/").run_async(&handler).await;
+                            if request.status().unwrap() == Status::Ok {
+                                Ok(())
+                            } else {
+                                Err(backoff::Error::transient(()))
+                            }
+                        }
+                    })
+                    .await
+                    .unwrap();
+                }
+            }));
+        }
+
+        debug!("waiting for queue to be full");
+        while get_outstanding_requests_gauge(metrics, meter_prefix).await
```

Comment:
Consider a `tokio::sleep` of a few milliseconds to avoid busy-looping here; it might be helpful to introduce a function since this loop is written multiple times. (Normally, I would suggest implementing a maximum wait time before failing, so that if this test breaks it won't silently hang until manually terminated -- but I think all of these loops are inside `RuntimeFlavor::run` which asserts a timeout of its own.)

Reply:
These loops shouldn't be truly busy loops, since the `.await` yields to the scheduler. I think of it as deprioritizing the test future as much as possible, until the condition is met. But, it may still be costly in terms of CPU, especially since we re-collect metrics on each loop iteration. I will refactor to a function and experiment with sleeps.

> but I think all of these loops are inside RuntimeFlavor::run which asserts a timeout of its own.

That's correct. As long as we haven't blocked the runtime itself, the test should cancel if we've broken the test. (Maybe I should do the thread-based timeout we mentioned in slack.)

Reply:
FWIW/IMO, I think the presence of an `.await` in a loop does not suffice to make it not-a-busy-loop. If no other tasks are schedulable (because there are no other tasks, or they are blocked waiting for something), this would still be a tight loop polling on a condition.

(The `.await` is definitely helpful, though -- I think it ensures that other tasks _can_ make progress if they are not blocked, even if we're using a single-threaded runtime.)